django/http/response.py

```diff
@@ -333,6 +334,9 @@ def __repr__(self):
             'content_type': self._content_type_for_repr,
         }
 
+    def __getstate__(self):
+        return {k: v for k, v in self.__dict__.items() if k not in self.non_pickable_attrs}
+
```

Comment:
I _think_ it's faster to do it this way:

```
def __getstate__(self):
    obj_dict = self.__dict__.copy()
    for attr in self.non_picklable_attrs:
        if attr in obj_dict:
            del obj_dict[attr]
        return obj_dict
```

which coincidentally is basically the same implementation as `SimpleTemplateResponse`

@felixxm Decision needed from you and/or anyone else appropriate. If `__getstate__` is implemented either as above, or as proposed by tigicion as of this revision (dict comprehension); do we want to adjust and simplify `SimpleTemplateReponse` and `TemplateResponse` as a corollary? The implementation for those would then be something like the following, I think:

```
class SimpleTemplateResponse(...):
    non_picklable_attrs = HttpResponse.non_picklable_attrs + ['template_name', 'context_data', '_post_render_callbacks']

    def __getstate__(self):
        if not self._is_rendered:
            raise ContentNotRenderedError('The response content must be rendered before it can be pickled.')
        return super().__getstate__()

class TemplateResponse(...):
    non_picklable_attrs = SimpleTemplateResponse.non_picklable_attrs + ['_request']
```
Though during the change I'd also propose changing the values to be `tuple` instead of `list` because mutable class defaults are nasty and presumably not actually desirable here.

We'd have to call out in the release notes that the attribute has changed name, at least.

Reply:
> @felixxm If `__getstate__` is implemented either as above, or as proposed by tigicion as of this revision (dict comprehension) and we want to adjust and simplify `SimpleTemplateReponse` and `TemplateResponse` as a corollary, would I be right in thinking the implementation for those would then be something like the following, before I send tigicion down the wrong path?

Agreed, the `rendering_attrs` attribute is undocumented, private API, so it's enough to document the name change in the release notes.

> Though during the change I'd also propose changing the values to be `tuple` instead of `list` because mutable class defaults are nasty and presumably not actually desirable here.

What do you think about using the `frozenset`?

Reply:
`frozenset` makes a lot of sense and doesn't get enough use IMHO, as long as we're careful to remember we need to use `|` rather than `+` on any subclasses that want to extend the data.